File: ocaml/plugins/secretspec.ml
(* The `secretspec` provider kind, as a voxgig/plugin definition.

   SecretSpec read through its own CLI - a child process, which is one of
   the three things that keeps a kind out of the core.

   A port of typescript/plugins/secretspec.ts, which is canonical. *)

open Secret
open Provider
open Runcmd

(* Does this SecretSpec failure mean "no such secret" rather than "I could
   not answer"?

   MATCHED ON THE WHOLE PHRASE, NOT ON "not found". SecretSpec says
   `Secret 'API_TOKEN' not found` for both a name it does not declare and
   one declared with no value, and both are misses. It also says
   `Provider backend 'keyring' not found`, which is a store that could not
   answer at all - and reading that as a miss is the worst failure this
   library has. The key is required to appear, so the two cannot be
   confused. *)
let secretspecmiss (why : string) (key : string) : bool =
  None <> findsub why ("Secret '" ^ key ^ "' not found")

(* SecretSpec (https://secretspec.dev), read through its CLI.

   `backend` selects one of SecretSpec's own backends (`--provider`) and is
   called `backend` here only because `provider` already means something
   else in this library.

   A reason is required, not optional: SecretSpec records every read in an
   audit log and refuses to read at all without one. *)
let secretspec_provider (commandgiven : string) (file : string) (profile : string)
    (backend : string) (reason : string) (prefix : string) : provider =
  let command = if "" = commandgiven then "secretspec" else commandgiven in

  {
    lookup =
      (fun name ->
        let key = envkey ~prefix name in

        (* `--file` before the subcommand, `--reason` always sent. *)
        let argv =
          [ command ]
          @ (if "" = file then [] else [ "--file"; file ])
          @ [ "get"; key ]
          @ (if "" = backend then [] else [ "--provider"; backend ])
          @ (if "" = profile then [] else [ "--profile"; profile ])
          @ [ "--reason"; firstof [ reason; "sekreto" ] ]
        in

        let ran = runcmd argv (Unix.environment ()) command in

        if 0 = ran.status then Some (dropsuffix ran.out "\n")
        else if secretspecmiss ran.why key then None
        else
          fail
            ("sekreto: secretspec error: "
            ^ if "" = ran.why then "exit " ^ string_of_int ran.status else ran.why));
    describe = (fun () -> if "" = backend then "secretspec" else "secretspec:" ^ backend);
  }

let plugin () : Defs.definition =
  providerplugin "secretspec" (fun spec ->
      secretspec_provider spec.command spec.file spec.profile spec.backend spec.reason spec.prefix)

File: c/plugins/secretspec.c
/* SecretSpec, through its CLI.
 *
 * A plugin because it spawns a child. `proc.c` is the only object in the
 * library that forks, and only this kind and `boru` pull it into a link.
 *
 * A port of typescript/plugins/secretspec.ts, which is canonical.
 */

#define _POSIX_C_SOURCE 200809L

#include <stdlib.h>
#include <string.h>

#include "support.h"

typedef struct {
  sek_pool *pool;
  const char *command;
  const char *file;
  const char *profile;
  const char *backend;
  const char *reason;
  const char *prefix;
  char *described;
} secretspecdata;

/* Does this SecretSpec failure mean "no such secret" rather than "I could
 * not answer"?
 *
 * MATCHED ON THE WHOLE PHRASE, NOT ON "not found". SecretSpec also says
 * `Provider backend 'keyring' not found`, which is a store that could not
 * answer at all - and reading that as a miss is the worst failure this
 * library has, because the chain then falls through to a weaker store
 * without saying so. The key is required to appear, so the two cannot be
 * confused. */
static int secretspecmiss(sek_pool *pool, const char *why, const char *key) {
  return sek_contains(why, sek_fmt(pool, "Secret '%s' not found", key));
}

static sek_err secretspec_lookup(sek_provider *self, const char *name, char **out) {
  secretspecdata *data = (secretspecdata *)self->data;
  sek_pool *pool = data->pool;
  char *key = NULL;
  sek_err err = sek_envkey(pool, name, data->prefix, &key);
  char *argv[12];
  size_t at = 0;
  sek_ran ran;
  size_t len;

  *out = NULL;

  if (NULL != err) {
    return err;
  }

  /* The order is exact: --file comes BEFORE the subcommand, and --reason
   * is always sent, because SecretSpec audits every read and refuses
   * without one. */
  argv[at++] = (char *)data->command;
  if (!sek_empty(data->file)) {
    argv[at++] = (char *)"--file";
    argv[at++] = (char *)data->file;
  }
  argv[at++] = (char *)"get";
  argv[at++] = key;
  if (!sek_empty(data->backend)) {
    argv[at++] = (char *)"--provider";
    argv[at++] = (char *)data->backend;
  }
  if (!sek_empty(data->profile)) {
    argv[at++] = (char *)"--profile";
    argv[at++] = (char *)data->profile;
  }
  argv[at++] = (char *)"--reason";
  argv[at++] = (char *)sek_first(data->reason, "sekreto");
  argv[at] = NULL;

  err = sek_runcmd(pool, argv, data->command, NULL, NULL, &ran);
  if (NULL != err) {
    return err;
  }

  if (0 == ran.status) {
    len = strlen(ran.out);
    *out = 1 <= len && '\n' == ran.out[len - 1] ? sek_strndup(pool, ran.out, len - 1) : ran.out;
    return NULL;
  }

  if (secretspecmiss(pool, ran.why, key)) {
    return NULL;
  }

  return sek_fmt(pool, "sekreto: secretspec error: %s",
                 sek_empty(ran.why) ? sek_fmt(pool, "exit %d", ran.status) : ran.why);
}

static const char *secretspec_describe(sek_provider *self) {
  return ((secretspecdata *)self->data)->described;
}

/* ---- the definition ------------------------------------------------ */

static sek_err secretspec_make(sek_pool *pool, const sek_spec *spec, sek_provider **out) {
  secretspecdata *data = (secretspecdata *)sek_alloc(pool, sizeof(secretspecdata));
  data->pool = pool;
  data->command = sek_own(pool, sek_empty(spec->command) ? "secretspec" : spec->command);
  data->file = sek_own(pool, spec->file);
  data->profile = sek_own(pool, spec->profile);
  data->backend = sek_own(pool, spec->backend);
  data->reason = sek_own(pool, spec->reason);
  data->prefix = sek_own(pool, spec->prefix);
  data->described = sek_empty(spec->backend) ? sek_strdup(pool, "secretspec")
                                             : sek_fmt(pool, "secretspec:%s", spec->backend);
  *out = sek_provider_new(pool, secretspec_lookup, secretspec_describe, data);
  return NULL;
}

static sek_providerkind SECRETSPEC_KIND;

Definition *sek_plugin_secretspec(void) {
  return sek_providerplugin(&SECRETSPEC_KIND, "secretspec", secretspec_make);
}
